published_date/published_date.py

```diff
@@ -0,0 +1,109 @@
+            return None
+
+        if 'next' in results:
+            url_ = results['next']['url']
+            self.published_course_next_page_list.append(url_)
+            logger.info(f"Pagination size {len(self.published_course_next_page_list)} for published_at date")
+
+    def published_date_resp_parsing(self, response):
+
+        logger.info("published_date_resp_parsing Call")
+
+        if response is None:
+            logger.info(f"Published course date response is None ")
+            return
+
+        start_time = time.time()
+
+        logger.info(f"published courses date collected so far : {len(self.published_course_date)}")
+        status = response.result().status_code
+        published_date_found = False
+        if status != 200:
+            logger.info(f"Response not successful with status code {status} due to {response.result().text}")
+            return
+
+        try:
+            audit_events = json.loads(response.result().text)
+        except JSONDecodeError as e:
+            logger.info(f"Error in parsing the response {e.message}")
+            return
+
+        if not audit_events:
+            logger.info(f"Response for fetching published date is empty {audit_events}")
+            return
+
+        events = audit_events['events']
+
+        for event in events:
+            if event['event_type'] == 'published':
+                course_id = event['links']['course']
+                published_date_found = True
+                self.published_course_date.update({course_id: event['created_at']})
+                logger.info(f"Published Date {event['created_at']} for course {course_id}")
+                break
+        if not published_date_found:
+            self.get_next_page_url(response)
+
+        seconds = time.time() - start_time
+        str_time = time.strftime("%H:%M:%S", time.gmtime(seconds))
+        logger.debug(f"Parsing the published date took {str_time} ")
+        return
+
+    def get_published_course_date(self,course_ids, next_page_links=None):
+        logger.info("Starting of get_published_course_date call")
+        with FuturesSession(max_workers=self.workers) as session:
+            headers = {'Content-type': 'application/json', 'Authorization': 'Bearer ' + self.canvas_token[0]}
+            if next_page_links is not None:
+                logger.info("Going through Next page URL set")
+                responses = []
+                for next_page_link in next_page_links:
+                    response = session.get(next_page_link, headers=headers)
+                    self.published_course_next_page_list.remove(next_page_link)
+                    responses.append(response)
+            else:
+                logger.info("Initial Round of Fetching course published date")
+                responses = [session.get(f'{self.canvas_url[0]}/api/v1/audit/course/courses/{course_id}?per_page=100', headers=headers)
+                for course_id in course_ids]
```

Comment:
I'd probably refactor this for readability, probably making the url a variable on the previous line and then using it on this one.